File: 543-diameter-of-binary-tree/diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int answer;

public:
    int solve(TreeNode* root) {
        if (!root) return 0;
        int left = solve(root->left);
        int right = solve(root->right);
        this->answer = max(this->answer, left+right+1);
        return max(left, right) + 1;
    }

    int diameterOfBinaryTree(TreeNode* root) {
        answer = 0;
        solve(root);
        return answer-1;
    }
};

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:    
    def dfs(self, root: Optional[TreeNode]) -> int:
        # return longest path, but update self.mx as it the root

        if root.left is None and root.right is None:
            return 1

        left = 0 if root.left is None else self.dfs(root.left)
        right = 0 if root.right is None else self.dfs(root.right)

        as_root = left + right + 1
        if as_root > self.mx:
            self.mx = as_root

        return max(left, right) + 1

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.mx = 0

        res = self.dfs(root)

        print(self.mx)
        print(res)

        return max(self.mx, res) - 1
